server/src/main/java/org/elasticsearch/common/util/CuckooFilter.java

```diff
@@ -0,0 +1,481 @@
+        int alternateBucket = alternateIndex(bucket, fingerprint);
+
+        // check all entries for both buckets and the evicted slot
+        return hasFingerprint(bucket, fingerprint) || hasFingerprint(alternateBucket, fingerprint) || evictedFingerprint == fingerprint;
+    }
+
+    /**
+     * Return's true if any of the entries in the bucket contain the fingerprint
+     */
+    private boolean hasFingerprint(int bucket, long fingerprint) {
+        long[] values = new long[entriesPerBucket];
+        int offset = getOffset(bucket, 0);
+        data.get(offset, values, 0, entriesPerBucket);
+
+        return Arrays.stream(values).anyMatch(value -> value == fingerprint);
+    }
+
+    /**
+     * Add's the hash to the bucket or alternate bucket.  Returns true if the insertion was
+     * successful, false if the filter is saturated.
+     */
+    boolean add(MurmurHash3.Hash128 hash) {
+        // can only use 64 of 128 bytes unfortunately (32 for each bucket), simplest
+        // to just truncate h1 and h2 appropriately
+        int bucket = hashToIndex((int) hash.h1);
+        int fingerprint = fingerprint((int) hash.h2);
+        return mergeFingerprint(bucket, fingerprint);
+    }
+
+    /**
+     * Attempts to merge the fingerprint into the specified bucket or it's alternate bucket.
+     * Returns true if the insertion was successful, false if the filter is saturated.
+     *
+     * Expert-level API, use {@link CuckooFilter#add(MurmurHash3.Hash128)} to insert
+     * values into the filter
+     */
+    boolean mergeFingerprint(int bucket, int fingerprint) {
+        // If we already have an evicted fingerprint we are full, no need to try
+        if (evictedFingerprint != EMPTY) {
+            return false;
+        }
+
+        int alternateBucket = alternateIndex(bucket, fingerprint);
+        if (tryInsert(bucket, fingerprint) || tryInsert(alternateBucket, fingerprint)) {
+            count += 1;
+            return true;
+        }
+
+        for (int i = 0; i < MAX_EVICTIONS; i++) {
+            // overwrite our alternate bucket, and a random entry
+            int offset = getOffset(alternateBucket, rng.nextInt(entriesPerBucket - 1));
+            int oldFingerprint = (int) data.get(offset);
+            data.set(offset, fingerprint);
+
+            // replace details and start again
+            fingerprint = oldFingerprint;
+            bucket = alternateBucket;
+            alternateBucket = alternateIndex(bucket, fingerprint);
+
+            // Only try to insert into alternate bucket
+            if (tryInsert(alternateBucket, fingerprint)) {
+                count += 1;
+                return true;
+            }
+        }
+
+        // If we get this far, we failed to insert the value after MAX_EVICTION rounds,
+        // so cache the last evicted value (so we don't lose it) and signal we failed
+        evictedFingerprint = fingerprint;
+        return false;
```

Comment:
It looks like if we add an entry when the filter is saturated that it will still change the filter since it still runs and maintains the evictions it tries while it works out that the filter is full? I'm wondering if this property of the filter is intended and further if its considered ok?

Reply:
Oh actually this might be ok because of https://github.com/elastic/elasticsearch/pull/35718/files#diff-53984b5b7b1ac2bfe9cafb420e574923R196 which checks the evicted fingerprint. Is that intended to overcome the problem above?

Reply:
We chatted about this offline, but for posterity, the second comment is correct.  

After an item runs through a full eviction cycle, if we can't find a slot we need to save the item to a holding-area before returning false.  If we don't, we'll have accidentally kicked an item out and lose it.  We could theoretically put it back into the map by "undoing" the evictions... but that would be relatively expensive compared to a dedicated eviction variable.